store/src/java/com/zimbra/cs/ldap/unboundid/UBIDLdapFilterFactory.java

```diff
@@ -860,6 +863,60 @@ public ZLdapFilter accountsWithDocumentEditing() {
                         FILTER_ACCOUNTS_WITH_DOCUMENT_EDITING));
     }
 
+    @Override
+    public ZLdapFilter accountsByCosesAndFeatureCheck(List<String> cosIds, String ldapAttribute) {
+        List<Filter> filters = Lists.newArrayList();
+        for (String cosId : cosIds) {
+            filters.add(Filter.createEqualityFilter(Provisioning.A_zimbraCOSId, cosId));
+        }
+        filters.add(Filter.createNOTFilter(Filter.createPresenceFilter(Provisioning.A_zimbraCOSId)));
+        return new UBIDLdapFilter(
+                FilterId.ACCOUNTS_BY_COSES_AND_LDAP_FEATURE_CHECK,
+                Filter.createANDFilter(
+                        FILTER_ALL_NON_SYSTEM_ACCOUNTS,
+                        Filter.createNOTFilter(
+                            Filter.createORFilter(
+                                Filter.createEqualityFilter(ldapAttribute, LdapConstants.LDAP_TRUE),
+                                Filter.createEqualityFilter(ldapAttribute, LdapConstants.LDAP_FALSE))),
+                        Filter.createORFilter(filters)));
+    }
+
+    @Override
+    public ZLdapFilter accountsByCosAndFeatureCheck(String cosId, String ldapAttribute) {
+        Cos cos = null;
+        try {
+            cos = Provisioning.getInstance().get(Key.CosBy.id, cosId);
+        } catch (Exception e) {
+            ZimbraLog.system.warn("unable to fetch cos " + cosId);
+        }
+
+        List<Filter> filters = Lists.newArrayList();
+        filters.add(Filter.createEqualityFilter(Provisioning.A_zimbraCOSId, cosId));
+        // cos is default, need to add presence filter to consider all accounts with auto cos
+        if(null != cos && cos.isDefaultCos()) {
+            filters.add(Filter.createNOTFilter(Filter.createPresenceFilter(Provisioning.A_zimbraCOSId)));
+        }
+        return new UBIDLdapFilter(
+                FilterId.ACCOUNTS_BY_COS_AND_LDAP_FEATURE_CHECK,
+                Filter.createANDFilter(
+                        FILTER_ALL_NON_SYSTEM_ACCOUNTS,
+                        Filter.createNOTFilter(
+                                Filter.createORFilter(
+                                        Filter.createEqualityFilter(ldapAttribute, LdapConstants.LDAP_TRUE),
+                                        Filter.createEqualityFilter(ldapAttribute, LdapConstants.LDAP_FALSE))),
+                        Filter.createORFilter(filters)));
+    }
```

Comment:
Please create a  separate centralized `Filter` method for the common logic for creating the main LDAP filter which will centralize the filter creation, making the code `DRY (Don't Repeat Yourself)` and easier to maintain.

Reply:
Here we are already using combined filter wherever it is possible. For the Ldap check and cos check. the filters are dependent on the dynamic values for which we can't create centralized filters